algorithmic-strategies/dynamic-programming.md

```diff
@@ -0,0 +1,302 @@
+  'remembering stuff to save time later'"
+```
+
+## Example: Longest Common Subsequence
+
+The longest common subsequence problem takes two strings `str1` and `str2`, and returns the length of their longest common subsequence.
+
+A subsequence of a string is a new string generated from the original string with some characters (can be none) deleted without changing the relative order of the remaining characters. (eg, "ace" is a subsequence of "abcde" while "aec" is not). A common subsequence of two strings is a subsequence that is common to both strings.
+
+If there is no common subsequence, return 0.
+
+| `str1`    | `str2`      | Subsequence | Length (result) |
+| --------- | ----------- | ----------- | --------------- |
+| `"abcde"` | `"ace"`     | `"ace"`     | 3               |
+| `"abcde"` | `"aqzcrwe"` | `"ace"`     | 3               |
+| `"abc"`   | `"abc"`     | `"abc"`     | 3               |
+| `"abc"`   | `"def"`     | `""`        | 0               |
+| `"abc"`   | `""`        | `""`        | 0               |
+
+### !callout-info
+
+## From Genomics to Git Diffs
+
+Variations of the longest common subsequence problem power solutions for problems ranging from gene sequence comparisons to Git diffs!
+
+### !end-callout
+
+One approach we might take to solve this problem is at each position, to consider the current letter in each string and the remaining portion of each string. If the current characters match, they contribute one matched character count to our total length, plus however many matches there are in the remainders of the strings.
+
+But maybe we can get a better alignment (a longer subsequence) if we don't take the current pair of characters as part of the subsequence. What if we advanced the first string, looking for a better place to start the match. Or what if we advanced the second string? And what if this wasn't a match in the first place? We can advance both strings. For any of these cases, the current characters don't add anything to the overall length of the subsequence. The final length would be the maximum of those three options.
+
+And how do we find the maximum subsequence of the remaining strings? Why with recursion!
+
+![The call tree for finding the longest common subsequence for the strings "ace" and "ae". Because there are three branches at each level, it spreads very quickly. There are large repeated sections of the call tree.](../assets/algorithmic-strategies_dynamic-programming_lcs.png)  
+_Fig. If the recursive explosion of Fibonacci seemed bad, get a load of this! Crossed-out nodes indicate calls that result in a 0 length due to at least one of the input strings being empty. The numbers indicate the maximum value being returned back from a particular branch of execution. Repeated regions are outlined in yellow._
+
+The explosion of calls in this diagram puts Fibonacci to shame! But for small examples, like in the example table above, it can be manageable with a traditional recursive implementation.
```

Comment:
reworded to avoid traditional
```suggestion
The explosion of calls in this diagram puts Fibonacci to shame! But for small examples, like in the example table above, it can be manageable with a recursive implementation like the following:
```